src/main/java/org/apache/commons/crypto/cipher/OpenSsl.java

```diff
@@ -317,53 +350,43 @@ public int update(byte[] input, int inputOffset, int inputLen,
      *         bounded by the appropriate padding bytes
      */
-    public int doFinal(ByteBuffer output) throws ShortBufferException,
+    public int doFinal(ByteBuffer input, ByteBuffer output) throws ShortBufferException,
             IllegalBlockSizeException, BadPaddingException {
         checkState();
         Utils.checkArgument(output.isDirect(), "Direct buffer is required.");
-        int len = OpenSslNative.doFinal(context, output, output.position(),
-                output.remaining());
-        output.position(output.position() + len);
-        return len;
+
+        return opensslBlockCipher.doFinal(input, output);
     }
 
+
     /**
-     * Encrypts or decrypts data in a single-part operation, or finishes a
-     * multiple-part operation.
+     * Continues a multi-part update of the Additional Authentication
+     * Data (AAD).
+     * <p>
+     * Calls to this method provide AAD to the cipher when operating in
+     * modes such as AEAD (GCM).  If this cipher is operating in
+     * either GCM mode, all AAD must be supplied before beginning
+     * operations on the ciphertext (via the {@code update} and
+     * {@code doFinal} methods).
+     *
+     * @param aad the buffer containing the Additional Authentication Data
      *
-     * @param output the byte array for the result
-     * @param outputOffset the offset in output where the result is stored
-     * @return the number of bytes stored in output
-     * @throws ShortBufferException if the given output byte array is too small
-     *         to hold the result
-     * @throws BadPaddingException if this cipher is in decryption mode, and
-     *         (un)padding has been requested, but the decrypted data is not
-     *         bounded by the appropriate padding bytes
-     * @throws IllegalBlockSizeException if this cipher is a block cipher, no
-     *         padding has been requested (only in encryption mode), and the
-     *         total input length of the data processed by this cipher is not a
-     *         multiple of block size; or if this encryption algorithm is unable
-     *         to process the input data provided.
      */
-    public int doFinal(byte[] output, int outputOffset)
-            throws ShortBufferException, IllegalBlockSizeException,
-            BadPaddingException {
-        checkState();
-        return OpenSslNative.doFinalByteArray(context, output, outputOffset,
-                output.length - outputOffset);
+    public void updateAAD(byte[] aad) {
+        this.opensslBlockCipher.updateAAD(aad);
     }
 
+
     /** Forcibly clean the context. */
     public void clean() {
-        if (context != 0) {
-            OpenSslNative.clean(context);
-            context = 0;
+        if (opensslBlockCipher != null) {
+            opensslBlockCipher.clean();
         }
     }
 
     /** Checks whether context is initialized. */
     private void checkState() {
-        Utils.checkState(context != 0);
```

Comment:
1. behavior is not changed.  The context was moved to to the Cipher(OpenSslFeedbackCipher).
The cipher will check its own state.

2. Well, I am considering to remove OpenSsl.java.  From my point of view, now this layer(OpenSsl.java) seems redundant , we can move the logic(paramenters check) in OpenSsl.java to OpenSslCipher.  
Then,  the codebase will be clear:
* OpenSslCipher which implements the CryptoCipher interface which is exposed to the users.
* OpenSslFeedbackCipher and its sub classes (OpenSslGaloisCounterMode,OpenSslCommonMode) are private classes. they do the real work(encription/decription) for different modes.
* OpenSslNative: JNI class
Maybe we can do this later.